File: src/utils/saveManager.ts

import { GameState } from '../types';

const SAVE_KEY_PREFIX = 'uniCity_v3_slot_';

interface SetupData {
    name: string;
    type1: string;
    type2: string;
}

export interface SaveSlotInfo {
    slot: number;
    name: string;
    day: number;
    money: number;
    timestamp: number;
    universityLabel: string;
}

export const SaveManager = {
    save: (slot: number, gameState: GameState, setupData: SetupData) => {
        try {
            const data = {
                gameState,
                setupData,
                timestamp: Date.now()
            };
            localStorage.setItem(`${SAVE_KEY_PREFIX}${slot}`, JSON.stringify(data));
            return true;
        } catch (e) {
            console.error("Save failed", e);
            return false;
        }
    },

    load: (slot: number): { gameState: GameState, setupData: any } | null => {
        try {
            const raw = localStorage.getItem(`${SAVE_KEY_PREFIX}${slot}`);
            if (!raw) return null;
            const data = JSON.parse(raw);
            if (data && data.gameState) {
                // Ensure array/object fields exist to prevent crashes on legacy saves
                if (!data.gameState.incomingStudents) data.gameState.incomingStudents = [];
                if (!data.gameState.badges) data.gameState.badges = [];
                if (!data.gameState.studentList) data.gameState.studentList = [];
                if (!data.gameState.faculty) data.gameState.faculty = [];
                if (!data.gameState.colleges) data.gameState.colleges = [];
                if (!data.gameState.financeHistory) data.gameState.financeHistory = [];
                if (!data.gameState.financeSettings) {
                    data.gameState.financeSettings = {
                        tuitionFeePerYear: 10, totalDailyBudget: 100000,
                        allocationWeights: { research: 20, cafeteria: 20, maintenance: 20, student: 20, faculty: 20 }
                    };
                }
                if (!data.gameState.admissionPolicy) {
                    data.gameState.admissionPolicy = { totalTarget: 0, minScore: 500, collegeTargets: {}, lastPlanUpdateYear: 0 };
                }
                if (!data.gameState.publicity) {
                    data.gameState.publicity = { activeCampaigns: [], activeBuffs: [], ministryBonus: 30 };
                }
                if (!data.gameState.liaison) {
                    data.gameState.liaison = { missions: {}, activeMissions: [], cooldowns: {}, grantDaysLeft: 0, grantDailyAmount: 0 };
                }
                return data;
            }
            return null;
        } catch (e) {
            console.error("Load failed", e);
            return null;
        }
    },

    delete: (slot: number) => {
        localStorage.removeItem(`${SAVE_KEY_PREFIX}${slot}`);
    },

    getSlotInfo: (slot: number): SaveSlotInfo | null => {
        try {
            const raw = localStorage.getItem(`${SAVE_KEY_PREFIX}${slot}`);
            if (!raw) return null;
            const d = JSON.parse(raw);
            return d ? {
                slot,
                name: d.gameState.universityName,
                day: d.gameState.day,
                money: d.gameState.money,
                timestamp: d.timestamp,
                universityLabel: d.gameState.universityLabel || '未知'
            } : null;
        } catch {
            return null;
        }
    },

    getLatestSlot: (): number | null => {
        let latestTime = 0;
        let latestSlot = null;
        [1, 2, 3].forEach(slot => {
            const info = SaveManager.getSlotInfo(slot);
            if (info && info.timestamp > latestTime) {
                latestTime = info.timestamp;
                latestSlot = slot;
            }
        });
        return latestSlot;
    }
};

File: MyUniversity.swiftpm/Sources/SaveManager.swift
import Foundation

struct SaveSlotInfo: Identifiable {
    var id: Int { slot }
    var slot: Int
    var name: String
    var day: Int
    var money: Double
    var timestamp: Double
    var universityLabel: String
}

struct SaveData: Codable {
    var gameState: GameState
    var setupName: String
    var setupType1: UniType1
    var setupType2: UniType2
    var timestamp: Double
}

enum SaveManager {
    private static let prefix = "uniCity_v3_slot_"

    static func save(slot: Int, gameState: GameState, name: String, type1: UniType1, type2: UniType2) -> Bool {
        let data = SaveData(gameState: gameState, setupName: name, setupType1: type1, setupType2: type2, timestamp: Date().timeIntervalSince1970)
        do {
            let encoded = try JSONEncoder().encode(data)
            UserDefaults.standard.set(encoded, forKey: "\(prefix)\(slot)")
            return true
        } catch {
            print("Save failed: \(error)")
            return false
        }
    }

    static func load(slot: Int) -> SaveData? {
        guard let data = UserDefaults.standard.data(forKey: "\(prefix)\(slot)") else { return nil }
        do {
            var saveData = try JSONDecoder().decode(SaveData.self, from: data)
            // Migration
            if saveData.gameState.publicity.ministryBonus == 0 {
                saveData.gameState.publicity.ministryBonus = 30
            }
            return saveData
        } catch {
            print("Load failed: \(error)")
            return nil
        }
    }

    static func delete(slot: Int) {
        UserDefaults.standard.removeObject(forKey: "\(prefix)\(slot)")
    }

    static func getSlotInfo(slot: Int) -> SaveSlotInfo? {
        guard let saveData = load(slot: slot) else { return nil }
        return SaveSlotInfo(
            slot: slot,
            name: saveData.gameState.universityName,
            day: saveData.gameState.day,
            money: saveData.gameState.money,
            timestamp: saveData.timestamp,
            universityLabel: saveData.gameState.universityLabel
        )
    }

    static func getLatestSlot() -> Int? {
        var latestTime: Double = 0
        var latestSlot: Int?
        for slot in 1...3 {
            if let info = getSlotInfo(slot: slot), info.timestamp > latestTime {
                latestTime = info.timestamp
                latestSlot = slot
            }
        }
        return latestSlot
    }
}
